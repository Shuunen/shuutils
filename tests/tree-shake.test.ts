import { build as esbuild } from 'esbuild'
import path from 'path'
import { fileURLToPath } from 'url'
import { expect, it } from 'vitest'

const currentDirectory = path.dirname(fileURLToPath(import.meta.url))
const resolveDir = path.join(currentDirectory, '../dist') // eslint-disable-line unicorn/prevent-abbreviations

async function build (contents: string) {
  const result = await esbuild({
    bundle: true, // eslint-disable-line @typescript-eslint/naming-convention
    minify: false, // eslint-disable-line @typescript-eslint/naming-convention
    platform: 'node',
    format: 'esm',
    stdin: { contents, resolveDir },
    write: false, // eslint-disable-line @typescript-eslint/naming-convention
  })
  return {
    output: result.outputFiles[0]?.text,
    nbOutputFiles: result.outputFiles.length,
    warnings: result.warnings,
    errors: result.errors,
  }
}

function clearFileHash (content: string) {
  return content.replace(/(?<name>[a-z]{3,30})-[a-z\d]+(?<ext>\.js)/giu, '$<name>$<ext>')
}

it('tree-shake test A', async () => {
  const result = await build(`import { getRandomNumber } from '.'
  console.log('tree-shaking test A, only using getRandomNumber', getRandomNumber(1, 10))`)
  result.output = clearFileHash(result.output ?? '')
  expect(result).toMatchSnapshot()
})

it('tree-shake test B', async () => {
  const result = await build(`import { getRandomNumber, getRandomString } from '.'
  console.log('tree-shaking test B, using getRandomNumber and getRandomString', getRandomNumber(1, 10), getRandomString(10))`)
  result.output = clearFileHash(result.output ?? '')
  expect(result).toMatchSnapshot()
})

it('tree-shake test C', async () => {
  const result = await build(`import { readableTimeAgo } from '.'
  console.log('tree-shaking test C, only using readableTimeAgo', readableTimeAgo())`)
  result.output = clearFileHash(result.output ?? '')
  expect(result).toMatchSnapshot()
})
