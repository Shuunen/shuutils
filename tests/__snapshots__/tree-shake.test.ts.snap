// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`tree-shake test A 1`] = `
{
  "errors": [],
  "nbOutputFiles": 1,
  "output": "// dist/index.js
function getRandomNumber(min = 0, max = 100) {
  return Math.floor(Math.random() * (max - min + 1) + min);
}

// <stdin>
console.log("tree-shaking test A, only using getRandomNumber", getRandomNumber(1, 10));
",
  "warnings": [],
}
`;

exports[`tree-shake test B 1`] = `
{
  "errors": [],
  "nbOutputFiles": 1,
  "output": "// dist/environment-C_L5l6Ng.js
function pickOne(items) {
  return items[Math.floor(Math.random() * items.length)];
}
function getRandomString() {
  const strings = "Bar Alto,Sin Seguritat,Lorem Ipsum,Ciao,Sit dolor,Por erestet,Tchu la Comida,Amet Inn,Aqualeris baked,Bouquet,Zu Amarillo,Ploject,Ruhe animals,Mah Plizure,Baacon pasty,Vinci mador,Alan Awake,Malohe Sutur,A priore sur,Quel memento,Kalitat arae".split(",");
  return String(pickOne(strings));
}

// dist/index.js
function getRandomNumber(min = 0, max = 100) {
  return Math.floor(Math.random() * (max - min + 1) + min);
}

// <stdin>
console.log("tree-shaking test B, using getRandomNumber and getRandomString", getRandomNumber(1, 10), getRandomString(10));
",
  "warnings": [],
}
`;

exports[`tree-shake test C 1`] = `
{
  "errors": [],
  "nbOutputFiles": 1,
  "output": "// dist/environment-C_L5l6Ng.js
var nbMsInSecond = 1e3;
var nbMsInMinute = 6e4;
var nbMsInHour = 36e5;
var nbMsInDay = 864e5;
var nbMsInWeek = 6048e5;
var nbMsInMonth = 2592e6;
var nbMsInYear = 31536e6;

// dist/index.js
var timeAgoTuples = [
  [nbMsInMinute, nbMsInSecond, "second"],
  [nbMsInHour, nbMsInMinute, "minute"],
  [nbMsInDay, nbMsInHour, "hour"],
  [nbMsInWeek, nbMsInDay, "day"],
  [nbMsInMonth, nbMsInWeek, "week"],
  [nbMsInYear, nbMsInMonth, "month"]
];
function readableTimeAgo(input, language = "en") {
  const rtf = new Intl.RelativeTimeFormat(language, { numeric: "auto" });
  const ms = typeof input === "number" ? input : Date.now() - input.getTime();
  for (const [threshold, divisor, unit] of timeAgoTuples)
    if (ms < threshold)
      return rtf.format(-Math.floor(ms / divisor), unit);
  return rtf.format(-Math.floor(ms / nbMsInYear), "year");
}

// <stdin>
console.log("tree-shaking test C, only using readableTimeAgo", readableTimeAgo());
",
  "warnings": [],
}
`;
